pywavefront/parser.py

```diff
@@ -31,44 +31,136 @@
 
-    def read_file(self):
-        with open(self.file_name, 'r') as file:
-            for line in file:
-                self.parse(line)
-
-    def parse(self, line):
-        """Determine what type of line we are and dispatch
-        appropriately."""
-        if line.startswith('#'):
-            return
-
-        values = line.split()
-        if len(values) < 2:
-            return
-
-        line_type = values[0]
-        args = values[1:]
-        attrib = 'parse_%s' % line_type
-
-        if Parser.strict:
-            parse_function = getattr(self, attrib)
-            parse_function(args)
-        elif hasattr(self, attrib):
-            parse_function = getattr(self, attrib)
-            parse_function(args)
+        self.dispatcher = self._build_dispatch_map()
+        self.lines = self.create_line_generator()
+
+        self.line = None
+        self.values = None
+
+    def create_line_generator(self):
+        """
+        Creates a generator function yielding lines in the file
+        Should only yield non-empty lines
+        """
+        if self.file_name.endswith(".gz"):
+            if sys.version_info.major == 3:
+                gz = gzip.open(self.file_name, mode='rt', encoding=self.encoding)
+                f = io.BufferedReader(gz)
+                for line in gz:
+                    if line == '':
+                        continue
+                    yield line
+
+                gz.close()
+            else:
+                gz = gzip.open(self.file_name, mode='rt')
+                f = io.BufferedReader(gz)
+                for line in f.readlines():
+                    if line == '':
+                        continue
+                    yield line
+
+                gz.close()
+        else:
+            if sys.version_info.major == 3:
+                # Python 3 native `open` is much faster
+                with open(self.file_name, mode='r', encoding=self.encoding) as file:
+                    for line in file:
+                        # Skip empty lines
+                        if line == '':
+                            continue
+                        yield line
+            else:
+                with codecs.open(self.file_name, mode='r', encoding=self.encoding) as file:
+                    for line in file:
+                        # Skip empty lines
+                        if line == '':
+                            continue
+                        yield line
+
+    def next_line(self):
+        """Read the next line from the line generator and split it"""
+        self.line = next(self.lines)
+        self.values = self.line.split()
+
+    def consume_line(self):
+        """Tell the parser we are done with this line"""
+        self.line = None
+        self.values = None
+
+    def parse(self):
+        """
+        Parse all the lines in the obj file
+        Determines what type of line we are and dispatch appropriately.
+        """
+        try:
+            while True:
```

Comment:
Where is the exit condition? I see a `continue`, but no `break`.

Reply:
Added more comments about this. The loop breaks when `StopIteration` is raised. This triggers from `next_line()` when the line generator has nothing more to give us.